tests/tools/fio/workload.go

```diff
@@ -42,7 +45,156 @@ func (fr *Runner) writeFiles(fullPath string, opt Options) error {
 
+// WriteFilesAtDepth writes files to a directory "depth" layers deep below
+// the base data directory.
+func (fr *Runner) WriteFilesAtDepth(relBasePath string, depth int, opt Options) error {
+	fullBasePath := filepath.Join(fr.LocalDataDir, relBasePath)
+
+	err := os.MkdirAll(fullBasePath, 0700)
+	if err != nil {
+		return errors.Wrapf(err, "unable to make base dir %v for writing at depth", fullBasePath)
+	}
+
+	return fr.writeFilesAtDepth(fullBasePath, depth, depth, opt)
+}
+
+// WriteFilesAtDepthRandomBranch writes files to a directory "depth" layers deep below
+// the base data directory and branches at a random depth.
+func (fr *Runner) WriteFilesAtDepthRandomBranch(relBasePath string, depth int, opt Options) error {
+	fullBasePath := filepath.Join(fr.LocalDataDir, relBasePath)
+
+	err := os.MkdirAll(fullBasePath, 0700)
+	if err != nil {
+		return errors.Wrapf(err, "unable to make base dir %v for writing at depth with a branch", fullBasePath)
+	}
+
+	return fr.writeFilesAtDepth(fullBasePath, depth, rand.Intn(depth+1), opt) // nolint:gosec
+}
+
 // DeleteRelDir deletes a relative directory in the runner's data directory.
 func (fr *Runner) DeleteRelDir(relDirPath string) error {
 	return os.RemoveAll(filepath.Join(fr.LocalDataDir, relDirPath))
 }
+
+// DeleteDirAtDepth deletes a random directory at the given depth.
+func (fr *Runner) DeleteDirAtDepth(relBasePath string, depth int) error {
+	if depth == 0 {
+		return ErrCanNotDeleteRoot
+	}
+
+	fullBasePath := filepath.Join(fr.LocalDataDir, relBasePath)
+
+	return fr.operateAtDepth(fullBasePath, depth, os.RemoveAll)
+}
+
+// DeleteContentsAtDepth deletes some or all of the contents of a directory
+// at the provided depths.
+func (fr *Runner) DeleteContentsAtDepth(relBasePath string, depth, pcnt int) error {
+	fullBasePath := filepath.Join(fr.LocalDataDir, relBasePath)
+
+	return fr.operateAtDepth(fullBasePath, depth, func(dirPath string) error {
+		fileInfoList, err := ioutil.ReadDir(dirPath)
+		if err != nil {
+			return err
+		}
+
+		for _, fi := range fileInfoList {
+			const hundred = 100
+			if rand.Intn(hundred) < pcnt { // nolint:gosec
+				path := filepath.Join(dirPath, fi.Name())
+				err = os.RemoveAll(path)
+				if err != nil {
+					return err
+				}
+			}
+		}
+
+		return nil
+	})
+}
+
+// List of known errors.
+var (
+	ErrNoDirFound       = errors.New("no directory found at this depth")
+	ErrCanNotDeleteRoot = errors.New("can not delete root directory")
+)
+
+func (fr *Runner) operateAtDepth(path string, depth int, f func(string) error) error {
+	if depth <= 0 {
+		log.Printf("performing operation on directory %s\n", path)
+		return f(path)
+	}
+
+	fileInfoList, err := ioutil.ReadDir(path)
+	if err != nil {
+		return errors.Wrapf(err, "unable to read dir at path %v", path)
+	}
+
+	var dirList []string
+
+	for _, fi := range fileInfoList {
+		if fi.IsDir() {
+			dirList = append(dirList, filepath.Join(path, fi.Name()))
+		}
+	}
+
+	rand.Shuffle(len(dirList), func(i, j int) {
```

Comment:
I think I understand what your concern is, correct me if I'm wrong.
Yes we'll do an exhaustive search in a given branch, if only a certain subset of directories has the requested depth to conduct the operation.
But if I call, e.g., `DeleteDirAtDepth("/", 4)` 2 times in a row, I don't want to be limited to only deleting from the initial alphabetical subdirectory `/aaaa/aaaa/aaaa/delete_this_directory_first`, then `/aaaa/aaaa/aaaa/delete_this_directory_second`. I actually want a chance of deleting `/zzzz/zzzz/zzzz/delete_this_too_by_chance`. Since it's recursive, I'll shuffle the subdirectory traversal pick at any given depth so I give a chance to other viable paths that can complete the operation.

Reply:
I'm confused.
Is `operateAtDepth` expected to execute `f` on a single directory a depth `depth`? or on all directories that happen to be found at the specified depth?
Should the loop below break, actually return, when `err == nil` ?
Or is the comment above regarding the case when an error is encountered and then the operation is repeated?

Reply:
`operateAtDepth` is "keep traversing into deeper subdirectories, picking a random subdirectory each time. Then when you pick a subdirectory that is at the requested depth, pass the path to that subdirectory into the function `f` to do whatever's needed. For example Deleting a directory at that depth is just `f = os.RemoveAll`. Deleting the contents of a directory at that depth but leaving the directory itself involves reading all of the contents and calling `os.RemoveAll`.

We only want to choose among paths that are as deep as the requested depth, so iterating through the subdirectories is only performed when a given branch's depth is not sufficient, which is what I was referring to in the comment above.

```
/ab
/aa/aa/aa/delete_this_directory
```
If it's looking for depth 4, it might pick `/ab` first, then when it doesn't find any subdirectories to further traverse, pops with `ErrNoDirFound`, in which case the parent loop will iterate within that parent directory and try `/aa`. That path will end up with a successful operation at depth 4.

Reply:
OK, but what are the expected semantics?

In the example above with 3 directories at the same level:

* `/aaaa/aaaa/aaaa/delete_this_directory_first`
* `/aaaa/aaaa/aaaa/delete_this_directory_second`
* `/zzzz/zzzz/zzzz/delete_this_too_by_chance`

If `DeleteDirAtDepth("/", 4)` is called, does a single (random) directory get removed? or do all of them get removed?

Reply:
The expectation is a single random directory. As soon as a successful operation takes place (e.g. `os.RemoveAll`), the `operateAtDepth` stack unwinds, returning `err == nil` (or any other error, as long as it's not `ErrNoDirFound`, which only happens if it couldn't find a directory to execute the operation on)